### ISSUE TITLE
List/Stream implementations visibility

### ISSUE THREAD

**Issue by ruslansennov** [javaslang/javaslang#945]
@romildo [says](https://gitter.im/javaslang/javaslang/archives/2015/12/10)

> The best way of constructing a non-empty list given its head h and its tail l is "new List.Cons(h, t)"?

Maybe we should hide <code>Cons</code> and <code>Nil</code> from public?

**Comment by danieldietrich** [comment]
I originally wanted it for pattern matching

```java
Match.of(list)
    .when(List(x, xs)).then(println("head: " + x + ", tail: " + xs))
    .otherwise(println("empty list"));
```

What other benefits could we have?

Ad long there is no decomposition in pattern matching I think we don't need them. Scala had them (for List only?).

**Comment by danieldietrich** [comment]
Are there other types that have inner classes we don't want to expose?

**Comment by danieldietrich** [comment]
I'm not sure if cyclops (@johnmcclean-aol) or derive4j (@jbgi) need access to Cons and Nil.

Does javaslang-jackson currently use them?

**Comment by ruslansennov** [comment]
> Are there other types that have inner classes we don't want to expose?

IMO all implementations of these interfaces should be hidden:
* List
* Stream
* Option (in case if Option.of(null) will return Some(null))
* Lazy (already hidden)

> Does javaslang-jackson currently use them?

Of course no! :)

**Comment by danieldietrich** [comment]
Basically we need the types for pattern matching, therefore they should stay with us.

```java
Supplier<R> s1 = Match.of(option)
    .whenType(List.Cons.class).then(some -> ...)
    .whenType(List.Nil.class).then(nil -> ...);

Supplier<R> s2 = Match.of(_try)
    .whenType(Success.class).then(success -> ...)
    .whenType(Failure.class).then(failure -> ...);
```

We could make the constructors (package) private to make clear how to obtain classes. There have to be factory methods for the subtypes. I began already with that. Example:

```java
<T> Option<T> of(T t);
```

**Comment by danieldietrich** [comment]
* Move impl classes into their interfaces
* Hide constructors of impl classes
* Ensure that special factory methods for impl classes exist (but they return the interface)

**Comment by ruslansennov** [comment]
Fixed by ff434724c11329d5e16e19eb7dbaf4aa90d5c5b7
